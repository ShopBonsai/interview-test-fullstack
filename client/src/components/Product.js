import React, { useState } from 'react';
import { CardTitle, CardSubtitle, CardText, CardBody, Media } from 'reactstrap';
import { ProductImage, AddToCart, QuantityInput } from './';

const Product = ({ product, updateCart }) => {
	const [selectedQuantity, setSelectedQuantity] = useState(1);
	const { color, description, image, name, price, size, id, quantity } =
		product;

	return (
		<Media key={id} className='product-card'>
			<ProductImage image={image} />
			<CardBody>
				<CardTitle style={{ fontWeight: 600 }}>{name}</CardTitle>
				<CardTitle>Price: {price}</CardTitle>
				<CardSubtitle>Color: {color}</CardSubtitle>
				<CardSubtitle>Size: {size}</CardSubtitle>
				<CardText>Details: {description}</CardText>
				<AddToCart
					quantity={selectedQuantity}
					productId={id}
					quantity={selectedQuantity}
					updateCart={updateCart}
				>
					<QuantityInput
						availableQuantity={quantity}
						selectedQuantity={selectedQuantity}
						setSelectedQuantity={setSelectedQuantity}
					/>
				</AddToCart>
			</CardBody>
		</Media>
	);
};

export default Product;
