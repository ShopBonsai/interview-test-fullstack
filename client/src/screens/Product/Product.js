import React from 'react';
import { useParams, Link } from 'react-router-dom';
import { fetchGet } from '../../helper/fetchHelper';
import {
  CardTitle,
  CardSubtitle,
  CardText,
  Button,
  CardBody,
  Media,
  Spinner,
} from 'reactstrap';

const Product = () => {
  const { id } = useParams();

  const [product, setProduct] = React.useState(null);

  React.useEffect(() => {
    const fetchProduct = async () => {
      const res = await fetchGet(`/products/${id}`);
      setProduct(res.product);
    };
    fetchProduct();
  }, []);

  return (
    <>
      {product === null ? (
        <Spinner color="primary" />
      ) : (
        <Media className="product-card">
          <Media left href="#">
            <Media object src={product.image_url} alt="Product image cap" />
          </Media>
          <CardBody>
            <CardTitle style={{ fontWeight: 600 }}>
              {product.product_name}
            </CardTitle>
            <CardTitle>Price: {product.price}</CardTitle>
            <CardSubtitle>Color: {product.color}</CardSubtitle>
            <CardSubtitle>Size: {product.size}</CardSubtitle>
            <CardText>Details: {product.description}</CardText>
            <Button color="primary" size="lg" block>
              Buy
            </Button>
          </CardBody>
        </Media>
      )}

      <Link to="/">
        <Button color="secondary" size="lg" block>
          Back
        </Button>
      </Link>
    </>
  );
};

export default Product;
