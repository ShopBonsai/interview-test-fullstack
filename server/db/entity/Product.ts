import { Field, Float, Int, ObjectType } from "type-graphql";
import { BaseEntity, Column, Entity, PrimaryGeneratedColumn } from "typeorm";

export interface IProduct {
  id: string;
  name: string;
  price: number;
  description: string;
  color: string;
  size: string;
  quantity: number;
  image: string;
  belongsToBrand: number;
}

@Entity()
@ObjectType()
export default class Product extends BaseEntity implements IProduct {
  @PrimaryGeneratedColumn('uuid')
  @Field()
  id!: string;

  @Column()
  @Field()
  name!: string;

  @Column()
  @Field(() => Float)
  price!: number;

  @Column()
  @Field()
  description!: string;

  @Column()
  @Field()
  color!: string;

  @Column()
  @Field()
  size!: string;

  @Column()
  @Field(() => Int)
  quantity!: number;

  @Column()
  @Field()
  image!: string;

  @Column()
  @Field(() => Int)
  belongsToBrand!: number;
}
