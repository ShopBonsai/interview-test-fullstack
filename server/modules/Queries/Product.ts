// ðŸ˜  It's called db3 because `Cannot redeclare block-scoped variable 'db'.`
// I do not have time to fix that rn
const db3 = require('../../db/connect');

/*
 * No longer using this to folder to query (from client)
 * produceViewProductEvent is now going be used in the /server.ts
 */
// const { produceViewProductEvent } = require('./viewProductProducer.js');
interface argsFindProduct {
  id: number;
  user_id?: string;
}

const FIND_PRODUCT = async (_: any, { id, user_id }: argsFindProduct) => {
  console.log('user_id', user_id);
  // Do this asynchronously
  // produceViewProductEvent({ id, user_id });
  return await db3('products').where('id', id).first();
};

module.exports = { FIND_PRODUCT };
