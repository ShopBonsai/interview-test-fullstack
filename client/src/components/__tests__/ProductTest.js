import React from 'react';
import Product from '../Product';
import ShallowRenderer from 'react-test-renderer/shallow';

describe('Product Component', () => {
    const addItemFn = jest.fn();
    const testProps = {
        addItem: addItemFn,
        product: {
            belongsToBrand: 2,
            id: '37b4d1b7-ed3c-4d28-91e0-ef0fb89b2b12',
            name: 'VENIAM Swimming Shorts',
            price: 706.2,
            description: 'Excepteur non Lorem aliqua esse irure veniam elit labore culpa velit nostrud ad nostrud.',
            color: 'deserunt',
            size: 'L',
            quantity: 1,
            image: 'my_test_image.png'
        }
    };
    it('renders correctly', () => {
        const renderer = new ShallowRenderer()
        renderer.render(<Product product={testProps.product} addItem={addItemFn} />);
        const renderedOutput = renderer.getRenderOutput();
        expect(renderedOutput.type).toBe('div');
        expect(renderedOutput).toMatchSnapshot();
    });
})

