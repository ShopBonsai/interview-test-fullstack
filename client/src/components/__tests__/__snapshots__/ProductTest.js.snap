// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Product Component renders correctly 1`] = `
<div>
  <Media
    className="product-card"
  >
    <Media
      href="#"
      left={true}
    >
      <Media
        alt="Product image cap"
        object={true}
        src="my_test_image.png"
      />
    </Media>
    <CardBody
      tag="div"
    >
      <CardTitle
        style={
          Object {
            "fontWeight": 600,
          }
        }
        tag="div"
      >
        VENIAM Swimming Shorts
      </CardTitle>
      <CardTitle
        tag="div"
      >
        Price: $
        706.2
      </CardTitle>
      <CardSubtitle
        tag="div"
      >
        Color: 
        deserunt
      </CardSubtitle>
      <CardSubtitle
        tag="div"
      >
        Size: 
        L
      </CardSubtitle>
      <CardText
        tag="p"
      >
        Details: 
        Excepteur non Lorem aliqua esse irure veniam elit labore culpa velit nostrud ad nostrud.
      </CardText>
      <InputGroup
        style={
          Object {
            "marginBottom": "4px",
          }
        }
        tag="div"
      >
        <InputGroupAddon
          addonType="prepend"
          tag="div"
        >
          Quantity
        </InputGroupAddon>
        <Input
          onChange={[Function]}
          type="number"
        />
      </InputGroup>
      <Button
        block={true}
        color="primary"
        onClick={[Function]}
        size="lg"
        tag="button"
      >
        Buy
      </Button>
    </CardBody>
  </Media>
</div>
`;
